brian2/codegen/runtime/cython_rt/templates/spikegenerator.pyx

```diff
@@ -1,39 +1,45 @@
 {% extends 'common.pyx' %}
 
-{% block template_support_code %}
-cdef int _start_idx = 0
-cdef double _current_time = 0.0
-{% endblock %}
-
 {% block maincode %}
 
-    {# USES_VARIABLES {_spikespace, N, t, dt, neuron_index, spike_time } #}
-     
-    global _start_idx, _current_time
-    
-    # Check if we have been reset, if so reset the cache
-    if t<_current_time:
-        _start_idx = 0
-    
-    _current_time = t
+    {# USES_VARIABLES {_spikespace, N, t, dt, neuron_index, spike_time, period, _lastindex } #}
 
     # TODO: We don't deal with more than one spike per neuron yet

```

Comment:
Yeah, I had a look at this as well, the problem is that it has only been introduced by numpy 1.9 and I would prefer not to force everyone to upgrade just for this. It also was about ten times slower in my example, but you are of course right that my proposal isn't great for big maximum spike times.Here's another variant that combines `unique` (without `return_count`) with `bincount`, which doesn't have the problem of generating an intermediate array of the size of `max_t / dt` (it only generates intermediate on the order of the length of `timebins`), this is significantly slower (about 20 times) for the densely filled arrays I used previously, but maybe that's still fine for a single preparatory step:
```Python
_, indices = np.unique(timebins, return_index=True)
bincounts = np.diff(np.hstack([indices, len(timebins)]))
```